bfvmm/include/hve/arch/intel_x64/ept/intrinsics.h

```diff
@@ -0,0 +1,525 @@
+        constexpr const auto uc_eo = 0x04;
+        constexpr const auto uc_re = 0x05;
+        constexpr const auto uc_pt = 0x07;
+
+        constexpr const auto wc_tp = 0x08;
+        constexpr const auto wc_ro = 0x09;
+        constexpr const auto wc_rw = 0x0B;
+        constexpr const auto wc_eo = 0x0C;
+        constexpr const auto wc_re = 0x0D;
+        constexpr const auto wc_pt = 0x0F;
+
+        constexpr const auto wt_tp = 0x20;
+        constexpr const auto wt_ro = 0x21;
+        constexpr const auto wt_rw = 0x23;
+        constexpr const auto wt_eo = 0x24;
+        constexpr const auto wt_re = 0x25;
+        constexpr const auto wt_pt = 0x27;
+
+        constexpr const auto wp_tp = 0x28;
+        constexpr const auto wp_ro = 0x29;
+        constexpr const auto wp_rw = 0x2B;
+        constexpr const auto wp_eo = 0x2C;
+        constexpr const auto wp_re = 0x2D;
+        constexpr const auto wp_pt = 0x2F;
+
+        constexpr const auto wb_tp = 0x30;
+        constexpr const auto wb_ro = 0x31;
+        constexpr const auto wb_rw = 0x33;
+        constexpr const auto wb_eo = 0x34;
+        constexpr const auto wb_re = 0x35;
+        constexpr const auto wb_pt = 0x37;
+
+        inline auto get(epte_t &entry) noexcept
+        { return get_bits(entry, mask) >> from; }
+
+        inline void set(epte_t &entry, epte_value_t val) noexcept
+        { entry = set_bits(entry, mask, val << from); }
+    }
+
+    inline void trap_on_access(epte_t &entry) noexcept {
+        read_access::disable(entry);
+        write_access::disable(entry);
+        execute_access::disable(entry);
+    }
+
+    inline void pass_through_access(epte_t &entry) noexcept {
+        read_access::enable(entry);
+        write_access::enable(entry);
+        execute_access::enable(entry);
+    }
+
+    // An epte is considered "present" if any bits 2:0 or bit 10 is set
+    inline bool is_present(epte_t &entry) noexcept
+    { return entry & 0x407 ? true : false; }
+
+    // Return true if the given entry maps a page frame (is a leaf entry)
+    inline bool is_leaf_entry(epte_t &entry) noexcept
+    { return entry_type::is_enabled(entry) ? true : false; }
+
+    // Set all bits in the entry to 0
+    inline void clear(epte_t &entry) noexcept
+    { entry = 0; }
+
+    // Returns the hpa that the physical address bits of the given entry points
+    inline hpa_t hpa(epte_t & entry)
+    { return (phys_addr_bits::get(entry) << phys_addr_bits::from); }
+
+    // Set the physical address bits of the given entry to the given hpa
+    inline void set_hpa(epte_t & entry, hpa_t hpa)
+    { phys_addr_bits::set(entry, (hpa >> phys_addr_bits::from)); }
+
+}
+
+namespace gpa
+{
+    // Bits of a gpa that describe an offset into a pml4 table
+    namespace pml4_index
+    {
+        constexpr const auto from = 39U;
+        constexpr const auto size = 9U;
+        constexpr const auto mask = 0xff8000000000ULL;
+        constexpr const auto shift = 36U;
```

Comment:
The gpa intrinsics for fields that describe an offset into a page table have both a shift and a from that are different. "from" describes the least significant bit of the field (just like all the other intrinsics), while "shift" represents how far you need to shift that field to calculate the effective page table entry offset. Given the following example from the Intel manual:

> An EPT PML4E is selected using the physical address
defined as follows:
— Bits 63:52 are all 0.
— Bits 51:12 are from the EPTP.
— Bits 11:3 are bits 47:39 of the guest-physical address.
— Bits 2:0 are all 0.

from = 39 (the least significant bit of the offset in the guest physical address)
shift = 36 (because you need to shift bit 39 to bit 3 to represent the pml4e index)

Since I don't want the caller of these intrinsics to have to do something like ```auto pml4e_offset = gpa::pml4_index::get(gpa) << 3;```, I could see two options:

1. leave these intrinsics as they are now
2. split these into two intrinsics, one to get the index (the literal field from the gpa) and another to calculate the effective offset.

Reply:
Option #2 is better, although I might add that are existing intrinsics already require the user to shift by 3 to get the index. I am fine with a function for get_index that does this for you.